core/utils/thread_pool.hpp

```diff
@@ -20,31 +21,76 @@ namespace kagome {
    * Creates `io_context` and runs it on `thread_count` threads.
    */
   class ThreadPool final : NonCopyable, NonMovable {
+    enum struct State : uint32_t { kStopped = 0, kStarted };
+
    public:
     explicit ThreadPool(size_t thread_count)
         : io_context_{std::make_shared<boost::asio::io_context>()},
           work_guard_{io_context_->get_executor()} {
       BOOST_ASSERT(thread_count > 0);
+      stop();
+
       threads_.reserve(thread_count);
       for (size_t i = 0; i < thread_count; ++i) {
         threads_.emplace_back([io{io_context_}] { io->run(); });
       }
     }
 
+    explicit ThreadPool(std::shared_ptr<boost::asio::io_context> io_context)
+        : io_context_{std::move(io_context)}, work_guard_{}, threads_{} {
+      stop();
+    }
+
     ~ThreadPool() {
-      io_context_->stop();
-      for (auto &thread : threads_) {
-        thread.join();
+      if (!threads_.empty()) {
+        io_context_->stop();
+        for (auto &thread : threads_) {
+          thread.join();
+        }
+      }
+    }
+
+    void start() {
+      execution_state_.store(State::kStarted, std::memory_order_release);
+    }
+
+    void stop() {
+      execution_state_.store(State::kStopped, std::memory_order_release);
+    }
+
+    template <typename F>
+    void execute(F &&func) {
+      BOOST_ASSERT(io_context_);
+      if (State::kStarted == execution_state_.load(std::memory_order_acquire)) {
+        io_context_->post(std::forward<F>(func));
+      }
+    }
+
+    template <typename F, typename... Args>
+    auto reinvoke(F &&func, Args &&...args)
+        -> std::optional<std::tuple<Args...>> {
+      if (!isInCurrentThread()) {
+        execute([func(std::forward<F>(func)),
+                 tup{std::tuple<std::decay_t<Args>...>{
+                     std::forward<Args>(args)...}}]() mutable {
+          std::apply([&](auto &&...a) mutable { func(std::move(a)...); },
+                     std::move(tup));
+        });
+        return std::nullopt;
       }
+      return {std::make_tuple(std::forward<Args>(args)...)};
     }
 
-    const std::shared_ptr<boost::asio::io_context> &io_context() const {
-      return io_context_;
+    bool isInCurrentThread() const {
+      BOOST_ASSERT(io_context_);
+      return io_context_->get_executor().running_in_this_thread();
     }
 
    private:
+    std::atomic<State> execution_state_;
```

Comment:
Is it needed?
`io_context` has [`.stopped()`](https://www.boost.org/doc/libs/1_72_0/doc/html/boost_asio/reference/io_context/stopped.html)

```suggestion
    std::atomic<State> execution_state_ = State::kStopped;
```